components/region_cache_memory_engine/src/write_batch.rs

```diff
@@ -62,102 +66,114 @@ impl RangeCacheWriteBatch {
     }
 
     fn write_impl(&mut self, seq: u64) -> Result<()> {
-        self.buffer
-            .iter()
-            .map(|e| (e.cf.as_str(), e.encode(seq)))
-            .try_for_each(|(cf, (key, value))| (self.apply_cb)(cf, key, value))
+        let (engine, filtered_keys) = {
+            let core = self.engine.core().lock().unwrap();
+            (
+                core.engine().clone(),
+                self.buffer
+                    .iter()
+                    .filter(|&e| e.should_write_to_memory(core.range_manager()))
+                    .collect::<Vec<_>>(),
+            )
+        };
+        filtered_keys
```

Comment:
what if the range is just evicted before this line?

Reply:
shouldn't taking a look on core() (line 70 above) prevent that? should i take a lock elsewhere to protect access to range_manager.

Reply:
In this case those keys will still be written too memory, range_manager will prevent them from being read.  This is less efficient, but still correct; I can hold the lock on core when writing, but that will reduce write throughput. Another approach is to move from mutex to a read-write lock (or sharded lock) for core: perhaps I'll do this.

Reply:
Addressed this: switched to sharded read write lock from cross beam, using (relatively low-cost) read() lock during WriteBatch::write.

Reply:
Holding a lock (even the read/write lock) is too heavy for write, which blocks the snapshot acquisition.

Reply:
Although we will not read it due to the filter of range_manager, we don't have the logic to prevent a range load. That is say we cannot avoid the following situation:
we have cache range k1..k10
1. write_impl: should_write_to_memory for k1..k10
2. evict k1..k10
3. load  k1..k10
4. write_impl: write_to_memory

Reply:
We can have this steps to handle it:
1. get ranges to write
2. recording the ranges
3. write_to_memory
4. for ranges that are still valid, clear the range. for ranges evicted, schedule a task to delete the range and clear the range after that.

Reply:
What do youi mean by 'clear' the range?

Reply:
```
struct RangeManager {
    ...
    ranges_being_written: BtreeSet<CacheRange>,
}
```
So, clear just mean delete the range in the ranges_being_written.

Reply:
@tonyxuqqi thoughts on this approach? we can also re-compute this during garbage collection.

Reply:
> We can have these steps to handle it:
> 
> 1. get ranges to write
> 2. recording the ranges
> 3. write_to_memory
> 4. for ranges that are still valid, clear the range. for ranges evicted, schedule a task to delete the range and clear the range after that.
> 
> So, load a range should if the range is being written.

I worry that it's not more efficient than a simple read-write lock.  For read lock, actually it's pretty fast.